website/blog/2020-12-04-typedset.md

```diff
@@ -0,0 +1,53 @@
+
+The `OptimizedTypeSet` and improvements to the above mentioned functions are merged to master, and will be available from Presto 0.244.
+
+<!--truncate-->
+
+In this post we will look at the methods used in this improvement.
+
+## Avoid Using Internal BlockBuilder
+
+`TypedSet` has an internal `BlockBuilder` and appends each `Block` position being added to it, as well as an external `BlockBuilder` to construct the results. Block building using `BlockBuilder` is very costly and inefficient, especially when the memory growth was not handled properly. The `BlockBuilder` was needed to keep track of the elements being added, so that it can:
+
+1. Resolve hash table probing collision 
+2. Rehash. 
+
+However, both are actually not a problem. In all use cases, whole blocks (instead of several positions of a block) are added to the TypedSet, and the problem of resolving collisions can be resolved by keeping track of the blocks being added. Rehashing is not needed at all since we can know the max number of entries before creating a `TypedSet` for most use cases. 
+
+In the new design, we provided methods that add a whole `Block` with different set operations: 
+* union
+* intersect
+* except
+
+These methods take in a `Block` and remember the positions in the set using `SelectedPositions` objects. The operations can be applied on the same set multiple times, i.e. you can union two Block A and B, then intersect with Block C, then minus(except) Block D. This way the internal operations on the Block elements can be streamlined to more efficient loops. The memory consumption and allocation were also reduced because now we only need to remember the selected positions instead of building a whole new Block.
+
+## Avoid Computing The Hash Positions Multiple Times
+
+Calculating the hash positions took over 60% of these function costs. For some operations like set intersection, the previous implementation requires creating multiple `TypedSet`s and calculating the hash position multiple times. For example, in `array_intersect` function, it builds one `TypedSet` R, and based on the probe result on R, adds new elements to another `TypedSet` L. 
+
+![Remote Exchange](/img/blog/2020-12-04-OptimizedTypedSet/TypedSet.png)
+
+As shown in the above picture, the “Contains” and “Add” operation would both calculate the hash positions. But actually the hashPosition calculated in hashtable A can be reused by hashtable B if the size and hash functions are the same. This hash position indicates the starting position of the linear probing for both hash tables.
```

Comment:
Why does this switch from L and R to A and B, are these examples showing before and after of the same optimization? I don't really understand the example.

```suggestion
As shown in the above picture, the `Contains` and `Add` operation would both calculate the hash positions. But actually the hashPosition calculated in hash table A can be reused by hash table B if the size and hash functions are the same. This hash position indicates the starting position of the linear probing for both hash tables.
```

Reply:
@aweisberg I recreated the picture and rewrote this paragraph like this:

```
As shown in the above picture, the `contains` and `add` operation would both calculate the hash positions for hash table `R`
 and `L` respectively. This hash position indicates the starting position of the linear probing for both hash tables. But actually the
 hashPosition calculated in hash table `R` can be reused by hash table `L` if the hash table sizes and hash functions are the 
same. 

In the new `OptimizedTypedSet`, the first step is the same as the original implementation, in which the right side `Block` is 
added to the set first, and hash table `R` is constructed. Intersecting the set with the left side `Block` was done by creating a 
new hash table `L` of the same size (size precalculated) internally: it checks if the new elements are contained in hash table `R`,
 and if not, add them to hash table `L`. Unlike the original implementation, the `contains` and `add` operation share the same hash position calculation. When the new hash table `L` is constructed, the hash table `R` will be discarded. 
```
Does this sound easier to understand?